benchmarks/Microsoft.AspNetCore.Routing.Performance/EndpointMetadataCollectionBenchmark.cs

```diff
@@ -0,0 +1,127 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using BenchmarkDotNet.Attributes;
+
+namespace Microsoft.AspNetCore.Routing
+{
+    public class EndpointMetadataCollectionBenchmark
+    {
+        private object[] _items;
+        private EndpointMetadataCollection _collection;
+
+        [Params(3, 10, 25)]
+        public int Count { get; set; }
+
+        [GlobalSetup]
+        public void Setup()
+        {
+            var seeds = new Type[]
+            {
+                typeof(Metadata1),
+                typeof(Metadata2),
+                typeof(Metadata3),
+                typeof(Metadata4),
+                typeof(Metadata5),
+                typeof(Metadata6),
+                typeof(Metadata7),
+                typeof(Metadata8),
+                typeof(Metadata9),
+            };
+
+            _items = new object[Count];
+            for (var i = 0; i < _items.Length; i++)
+            {
+                _items[i] = seeds[i % seeds.Length];
+            }
+
+            _collection = new EndpointMetadataCollection(_items);
+        }
+
+        // This is a synthetic baseline that visits each item and does an as-cast.
+        [Benchmark(Baseline = true, OperationsPerInvoke = 5)]
+        public void Baseline()
+        {
+            var items = _items;
+            for (var i = items.Length - 1; i >= 0; i--)
+            {
+                GC.KeepAlive(_items[i] as IMetadata1);
```

Comment:
What does KeepAlive do?

Reply:
KeepAlive prevents the use of the argument from being optimized away. Basically it's a way to force the expression to be evaluated pretty much as written, without being deleted by the JIT.